Switch to representing FloatLiteralType as a RealId.

zygoloid opened the pull request:
Don't convert to f64 until we know that's the type that we actually
want. Also reimplement the conversion from RealId to FloatId to perform
an exact conversion with a real check for overflow, rather than
performing an approximate conversion via the host `double` type.

Unfortunately, LLVM doesn't expose its integer mantissa and exponent to
APFloat conversion, so we convert the RealId back to a string for now.

The LLVM conversion also detects overflow only if the literal would
round to having an out-of-range exponent, not if the literal is outside
the range of values of the type as the Carbon design expects. It's not
clear to me which rule we actually want here, so for simplicitly I'm
using the LLVM rule for now.

In preparation for adding other floating-point types beyond f64.

dwblaikie commented on toolchain/check/eval.cpp:
```diff
@@ -1071,6 +1072,76 @@ static auto PerformCheckedIntConvert(Context& context, SemIR::LocId loc_id,
+        result.convertFromString(str, llvm::APFloat::rmNearestTiesToEven);
+    if (auto error = status.takeError()) {
+      // The literal we create should always successfully parse.
+      CARBON_FATAL("Float literal parsing failed: {0}",
```
`CARBON_FATAL` always runs, yeah? (like it doesn't get omitted in non-debug builds) just checking 'cause otherwise `llvm::Error` would flag this as dropping an error on the floor

zygoloid replied:
Yeah, I think the plan is that this always remains fatal, even if we start turning off CHECKs in NDEBUG. It's possible that we'd stop producing the string in that case but I think we'd still arrange to mark it as "used".

dwblaikie commented on toolchain/check/testdata/basics/numeric_literals.carbon:
```diff
@@ -95,12 +139,12 @@ let e: f64 = 5.0e39999999999999999993;
-// CHECK:STDOUT:   %float.b98: Core.FloatLiteral = float_value 1.0E+8 [concrete]
-// CHECK:STDOUT:   %float.34d: Core.FloatLiteral = float_value 1.0E-8 [concrete]
+// CHECK:STDOUT:   %.7d0: Core.FloatLiteral = float_literal_value 9e-1 [concrete]
+// CHECK:STDOUT:   %.4bc: Core.FloatLiteral = float_literal_value 80e-1 [concrete]
+// CHECK:STDOUT:   %.24e: Core.FloatLiteral = float_literal_value 800e-1 [concrete]
```
Strange result - the 80e-1 and 800e-1 rather than 8e1 and 8e2? I guess this falls out of something else, so probably not worth changing - just odd.

zygoloid replied:
Yeah, this is a consequence of using `RealId` here -- that is represented as, and gets printed as, an integer value and an exponent, so there's never a decimal point.

dwblaikie reviewed (approved):
Why `float_literal_value` rather than `float_value` to match `int_value`?

And it looks like the instruction naming has had some loss, now it's often `%.XXX` instead of `%float.XXX`?

zygoloid committed:
Switch to representing FloatLiteralType as a RealId.

Don't convert to f64 until we know that's the type that we actually
want. Also reimplement the conversion from RealId to FloatId to perform
an exact conversion with a real check for overflow, rather than
performing an approximate conversion via the host `double` type.

Unfortunately, LLVM doesn't expose its integer mantissa and exponent to
APFloat conversion, so we convert the RealId back to a string for now.

The LLVM conversion also detects overflow only if the literal would
round to having an out-of-range exponent, not if the literal is outside
the range of values of the type as the Carbon design expects. It's not
clear to me which rule we actually want here, so for simplicitly I'm
using the LLVM rule for now.

zygoloid committed:
Add 'float' prefix in instnamer for float literals.

zygoloid reviewed (commented):
> Why `float_literal_value` rather than `float_value` to match `int_value`?

We use `float_value` for `FloatValue` (represented as a `FloatId` / an `APFloat`), so I'm intentionally using a different IR name for `FloatLiteralValue` (represented as a `RealId` / a mantissa + exponent as `APInt`s).

> And it looks like the instruction naming has had some loss, now it's often %.XXX instead of %float.XXX?

Yeah, that's not great. Fixed to use `%float.XXX` for `float_literal_value`s too.
